core/src/serialization.rs

```diff
@@ -0,0 +1,259 @@
+pub enum StringEncoding {
+    Canonical,
+    Indefinite(Vec<(u64, Sz)>),
+    Definite(Sz),
+}
+
+impl Default for StringEncoding {
+    fn default() -> Self {
+        Self::Canonical
+    }
+}
+
+impl From<cbor_event::StringLenSz> for StringEncoding {
+    fn from(len_sz: cbor_event::StringLenSz) -> Self {
+        match len_sz {
+            cbor_event::StringLenSz::Len(sz) => Self::Definite(sz),
+            cbor_event::StringLenSz::Indefinite(lens) => Self::Indefinite(lens),
+        }
+    }
+}
+
+#[inline]
+pub fn fit_sz(len: u64, sz: Option<cbor_event::Sz>, force_canonical: bool) -> Sz {
+    match sz {
+        Some(sz) => if !force_canonical && len <= sz_max(sz) {
+            sz
+        } else {
+            Sz::canonical(len)
+        },
+        None => Sz::canonical(len),
+    }
+}
+
+impl LenEncoding {
+    pub fn to_len_sz(&self, len: u64, force_canonical: bool) -> cbor_event::LenSz {
+        if force_canonical {
+            cbor_event::LenSz::Len(len, cbor_event::Sz::canonical(len))
+        } else {
+            match self {
+                Self::Canonical => cbor_event::LenSz::Len(len, cbor_event::Sz::canonical(len)),
+                Self::Definite(sz) => if sz_max(*sz) >= len {
+                    cbor_event::LenSz::Len(len, *sz)
+                } else {
+                    cbor_event::LenSz::Len(len, cbor_event::Sz::canonical(len))
+                },
+                Self::Indefinite => cbor_event::LenSz::Indefinite,
+            }
+        }
+    }
+
+    pub fn end<'a, W: Write + Sized>(&self, serializer: &'a mut Serializer<W>, force_canonical: bool) -> cbor_event::Result<&'a mut Serializer<W>> {
+        if !force_canonical && *self == Self::Indefinite {
+            serializer.write_special(cbor_event::Special::Break)?;
+        }
+        Ok(serializer)
+    }
+}
+
+impl StringEncoding {
+    pub fn to_str_len_sz(&self, len: u64, force_canonical: bool) -> cbor_event::StringLenSz {
+        if force_canonical {
+            cbor_event::StringLenSz::Len(cbor_event::Sz::canonical(len))
+        } else {
+            match self {
+                Self::Canonical => cbor_event::StringLenSz::Len(cbor_event::Sz::canonical(len)),
+                Self::Definite(sz) => if sz_max(*sz) >= len {
+                    cbor_event::StringLenSz::Len(*sz)
+                } else {
+                    cbor_event::StringLenSz::Len(cbor_event::Sz::canonical(len))
+                },
+                Self::Indefinite(lens) => cbor_event::StringLenSz::Indefinite(lens.clone()),
+            }
+        }
+    }
+}
+
+pub trait Serialize {
+    fn serialize<'a, W: Write + Sized>(
+        &self,
+        serializer: &'a mut Serializer<W>,
+        force_canonical: bool,
+    ) -> cbor_event::Result<&'a mut Serializer<W>>;
+
+    /// Bytes of a structure using the CBOR bytes as per the CDDL spec
+    /// which for foo = bytes will include the CBOR bytes type/len, etc.
+    fn to_cbor_bytes(&self, force_canonical: bool) -> Vec<u8> {
```

Comment:
One thing we have different is that CIP25 was generated without encoding preservation so there's this `force_canonical` concept on serialization. It's ignored by the auto-implementation of `Serialize` if `cbor_event::Serialize` is implemented, which is the trait that CIP25 uses. Is that the behavior we wanted? CIP25 was generated without the encoding preservation since it didn't really need it, from what I thought you wouldn't be storing the CBOR here and you'd be using it more for parsing existing metadata. The `force_canonical` is there since it's used for hardware wallets for encoding tx bodies but might not be too useful in other places.

Reply:
if it was generated with the encoding preservation, would this return just the original rep? canonical means https://www.rfc-editor.org/rfc/rfc7049#section-3.9 canonical cbor from the standard, right?

Reply:
So the stuff in `chain` right now if you pass in `false` will return whatever format was used when it was deserialized, or canonical if there was no format (e.g. was created from code right now not deserialized). If you pass in `true` it will turn it into canonical CBOR regardless of how it was created.

In CIP25 whatever you pass in it'll end up being canonical (afaik). So if you deserialize something it won't preserve the format, which was fine before when it was on its own since you're not going to be storing anything here, as this is more just for parsing without reencoding ever.

We could regenerate it with preserve-encdoings=true but that won't work with how it's used in Cardano as the CDDL file for the spec is actually a subset of the input bytes since it only parses the subset of a given tx metadata that conforms to the spec. But to make round-tripping consistent you'd need to figure out a way to remember the encodings of all other parts of the metadata NOT involved in the cip25 spec. We could potentially look at modifying cddl-codegen to figure out a way to do that but I'm not sure it's worth the time or complexity it adds to the crate. It would be nice if it was preserve-encodings for consistency now that we have a mono-repo and it's a part of CML, but on the other hand we could just document that it's not for round-tripping unless we think this is something important. @SebastienGllmt

Reply:
Yeah, handling this kind of subset operation is a ticket we have no plans to address at the moment

Reply:
So I was thinking about two options then:
a) Leave it how it is and document that the cip-25 crate doesn't preserve encodings and should not be used for round-tripping.
b) get rid of `Serialize::to_cbor_bytes(force_canonical: bool)` and instead make 2 other traits `SerializeOriginal::to_original_cbor_cbytes()` and `SerializeCanonical::to_canonical_cbor_bytes()` but it's feeling awfully pedantic now. In this case `cip-25` would only implement the canonical one and `chain` would implement both. Although this doesn't explicitly cover the fact that `cip-25` ignores fields not part of the spec either.

or something along those lines